src/Foundation/NSFastEnumerator.cs

```diff
@@ -43,15 +44,26 @@ void Fetch ()
 			count = NSFastEnumerator.objc_msgSend (collection.Handle, Selector.GetHandle ("countByEnumeratingWithState:objects:count:"), ref state, array, (nuint) array.Length);
 			if (!started) {
 				started = true;
-				mutationValue = Marshal.ReadIntPtr (state.mutationsPtr);
+				try {
+					// If Marshal.ReadIntPtr throws we fallback into using the pointer value instead.
```

Comment:
This is wrong from what I can tell, because the C# code is doing exactly the same as the equivalent Objective-C code (and if you're right, the Objective-C code would crash).

Note that the enumeration is implemented by clang, so this is not visible for normal developers.

Can you dump out the memory for the state (all 64 bytes) for the C# version and an identical Objective-C sample, to see if the Objective-C sample also gets the same values?

Reply:
@rolfbjarne You were right, the Objective-C sample crashes aswell with an `EXC_BAD_ACCESS (code=1, address=0x2)` asap it reaches the `for (AVCaptureOutput *dataOutput in synchronizedDataCollection)`. I used the following to inspect the internal `NSFastEnumerationState`:

```objc
- (void) inspectEnumerator:(AVCaptureSynchronizedDataCollection *)collection {
	// - (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state objects:(id __unsafe_unretained _Nullable [_Nonnull])buffer count:(NSUInteger)len;
	
	id (*my_msgSend)(id, SEL, NSFastEnumerationState *, id __unsafe_unretained *, NSUInteger) = (void *) objc_msgSend;
	NSFastEnumerationState state;
	NSUInteger len = 16;
	id __unsafe_unretained *buffer = (id __unsafe_unretained *) calloc (len, sizeof (id));
	
	NSUInteger count = (NSUInteger) my_msgSend (collection, @selector (countByEnumeratingWithState:objects:count:), &state, buffer, len);
	
	NSLog (@"%lu", (unsigned long)count);
	
	free (buffer);
}
```

and I get 

```lldb
(lldb) p state
(NSFastEnumerationState) $0 = {
  state = 1
  itemsPtr = 0x00000001c40f0688
  mutationsPtr = 0x0000000000000002
  extra = ([0] = 4373079224, [1] = 7583386624, [2] = 7583386624, [3] = 27017199717711872, [4] = 0)
}
```

So the crash makes sense since `mutationsPtr` is set to `0x2` and it is the same behavior I get in C# so I'll fill a radar. Thanks for noticing this @rolfbjarne 👍